src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -469,6 +482,31 @@ private void BuildFinished(object sender, BuildFinishedEventArgs e)
         _testEndTime = null;
     }
 
+    private void RenderBuildSummary()
+    {
+        if (!_projects.Any(p => p.Value.GetBuildErrorAndWarningMessages().Any()))
+        {
+            // No errors to display.
+            return;
+        }
+
+        Terminal.WriteLine(ResourceUtilities.GetResourceString("BuildSummary"));
+
+        foreach (Project project in _projects.Values.Where(p => p.GetBuildErrorAndWarningMessages().Any()))
+        {
+            string projectFileName = Path.GetFileNameWithoutExtension(project.File);
+            string? tfm = project.TargetFramework;
+            Terminal.WriteLine($"{Indentation}{projectFileName}{(tfm is null ? string.Empty : " ")}{AnsiCodes.Colorize(tfm, TerminalLogger.TargetFrameworkColor)}");
+
+            foreach (BuildMessage buildMessage in project.GetBuildErrorAndWarningMessages())
```

Comment:
Can we evaluate the performance impact of this please?
In the previous comment I asked about the `.GetBuildErrorAndWarningMessages()` iterating over large collection and here this is a second time we do this for every project.

Reply:
There is no I/O and all data is in memory. Also, there is only one full iteration per project as `Any()` just checks if the sequence is not empty. The summary is displayed after the build, so the performance of the build itself is not affected by this change.

Reply:
The underlying GetBuildErrorAndWarningMessages iterates over whole message collection, no matter the size - which happens both here and for the Any() check(to gather the Error/Warning messages to query the Any() against).
See my other comment.

Reply:
The underlying `GetBuildErrorAndWarningMessages` does not iterate; it creates an instance of the iterator. `Any` uses the iterator until it hits the first item in the sequence, so it could possibly iterate over the whole collection, but only if the error/warning is the last item or if there are no errors/warnings at all, which, I admit, is the desired state.

We could introduce a boolean field indicating if there is at least one error/warning in the project message list, but that means we need a condition that will check the severity of the message for each message emission, and this condition would be evaluated even in builds where the summary is disabled.

I tried to build OrchardCore (several times) with the summary enabled and having some NuGet errors. The logic took ~6ms on a build that took over a minute. I still don't think the performance is a concern here :)